### PR TITLE
refactor np.where to use overload

### PR DESCRIPTION
**Pull request by guilhermeleobas**
xref: #8254 

* Refactor `np.where` to use `@overload`
* Broadcast arrays of different shapes. This removes a restriction in the old code where only arrays with the same shape were supported.
* Add NumPy tests
* Add type check for supported types + tests

### CODE REVIEW

**Review by apmasell — DISMISSED** [review]

**Comment by guilhermeleobas** [comment]
Thanks for the review, @apmasell

**Comment by stuartarchibald** [comment]
@guilhermeleobas thanks for the patch. In the interests of making this easier to review, perhaps split the API refactoring and the addition of new functionality into separate patches?

**Comment by stuartarchibald** [comment]
@guilhermeleobas please could you resolve the conflicts against `main`? Many thanks.

File: numba/tests/test_array_methods.py [hunk 1]
```diff
@@ -916,6 +916,126 @@ def scalar_permutations():
...
+
+        # non bool mask
+        c = c.astype(int)
+        c[c != 0] = 34242324
+        np.testing.assert_equal(cfunc(c, a, b), r)
+        # invert
+        tmpmask = c != 0
+        c[c == 0] = 41247212
+        c[tmpmask] = 0
+        np.testing.assert_equal(cfunc(c, b, a), r)
```
Comment (kc611) [inline numba/tests/test_array_methods.py:996]: We could add some tests involving `None` ?

**Review by kc611 — CHANGES_REQUESTED** [review]
The code looks functionally good and well-written. There are mostly issues with handling of `None` and a few nitpicks but otherwise it LGTM.

File: numba/np/arraymath.py [hunk 2]
```diff
@@ -3255,120 +3253,102 @@ def array_nonzero(context, builder, sig, args):
...
+        special_0_case = all([check_0_dim(a) for a in (condition, x, y)])
+        if special_0_case:
+            return _where_zero_size_array_impl(dtype)
+
+        layout = condition.layout
+        if x_arr and y_arr:
+            if x.layout == y.layout == condition.layout:
+                layout = x.layout
+            else:
+                layout = 'C'
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/np/arraymath.py:3346]: I just kept the original behavior, which assigns the layout to `C`:
https://github.com/numba/numba/blob/3bee7be27b3672519a3e7660192dbb061ab424f4/numba/core/typing/npydecl.py#L710-L715

File: numba/tests/test_array_methods.py [hunk 3]
```diff
@@ -916,6 +916,126 @@ def scalar_permutations():
...
+            np.testing.assert_equal(cfunc(c[::3], d[::3], e[::3]), r[::3])
+            np.testing.assert_equal(cfunc(c[1::3], d[1::3], e[1::3]), r[1::3])
+            np.testing.assert_equal(cfunc(c[::-2], d[::-2], e[::-2]), r[::-2])
+            np.testing.assert_equal(cfunc(c[::-3], d[::-3], e[::-3]), r[::-3])
+            np.testing.assert_equal(cfunc(c[1::-3], d[1::-3], e[1::-3]), r[1::-3])
+
+    def test_np_where_numpy_ndim(self):
+        # https://github.com/numpy/numpy/blob/fe2bb380fd9a084b622ff3f00cb6f245e8c1a10e/numpy/core/tests/test_multiarray.py#L8737-L8749
+        pyfunc = np_where_3
+        cfunc = jit(nopython=True)(pyfunc)
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/tests/test_array_methods.py:955]: I'm following the convention used in the test file.

File: numba/np/arraymath.py [hunk 4]
```diff
@@ -3255,120 +3253,102 @@ def array_nonzero(context, builder, sig, args):
...
 
-    res = context.compile_internal(builder, scalar_where_impl, sig, args)
-    return impl_ret_new_ref(context, builder, sig.return_type, res)
+    for arg, name in zip((x, y), ('x', 'y')):
+        if not is_nonelike(arg) and not type_can_asarray(arg):
+            msg = 'The argument "{}" must be array-like if provided'
+            raise NumbaTypeError(msg.format(name))
+
+    if is_nonelike(x) and is_nonelike(y):
+        return _where_cond_none_none
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/np/arraymath.py:3332]: Supporting `None` inputs will be tricky. I'll try to address this in the following days.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, this largely looks good, great to see the NumPy tests passing too. I've left a few comments inline, once resolved should be good to merge.

File: numba/np/arraymath.py [hunk 5]
```diff
@@ -3255,120 +3253,112 @@ def array_nonzero(context, builder, sig, args):
...
-        np.where(scalar, scalar, scalar): return a 0-dim array
-        """
-        scal = x if cond else y
-        # This is the equivalent of np.full_like(scal, scal),
-        # for compatibility with Numpy < 1.8
-        arr = np.empty_like(scal)
-        arr[()] = scal
-        return arr
+    if is_nonelike(x) ^ is_nonelike(y):
+        msg = 'either both or neigher of "x" or "y" should be given'
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/np/arraymath.py:3313]: This is actually the message that NumPy uses:

```python
>>> import numpy as np
>>>a = np.arange(5)
>>> np.where(a, a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 180, in where
ValueError: either both or neither of x and y should be given
```

File: numba/np/arraymath.py [hunk 6]
```diff
@@ -3255,120 +3253,112 @@ def array_nonzero(context, builder, sig, args):
...
+        return x_ if condition else y_
+    return impl
 
 
 @register_jitable
-def _where_x_scalar(cond, x, y, res):
+def _where_generic_inner_impl(cond, x, y, res):
     for idx, c in np.ndenumerate(cond):
-        res[idx] = x if c else y[idx]
+        res[idx] = x[idx] if c else y[idx]
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/np/arraymath.py:3271]: Is there anything we can do in this case?

File: numba/np/arraymath.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/np/arraymath.py:3313]: Thanks for checking, I guess we should leave it then.

File: numba/np/arraymath.py [hunk 7]
```diff
@@ -3205,120 +3203,102 @@ def array_nonzero(context, builder, sig, args):
...
 
-    res = context.compile_internal(builder, scalar_where_impl, sig, args)
-    return impl_ret_new_ref(context, builder, sig.return_type, res)
+    for arg, name in zip((x, y), ('x', 'y')):
+        if not is_nonelike(arg) and not type_can_asarray(arg):
+            msg = 'The argument "{}" must be array-like if provided'
+            raise NumbaTypeError(msg.format(name))
+
+    if is_nonelike(x) and is_nonelike(y):
+        return _where_cond_none_none
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/np/arraymath.py]: I've changed the code to not use `None` as default value.

**Comment by guilhermeleobas** [comment]
@kc611 @stuartarchibald, would you folks be ok with not supporting `None` inputs in this patch? Current patch fails with such values as `np.asarray(None)` is not supported.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the updates @guilhermeleobas, think they address everything in the review. I'm inclined to leave the issue with unifying 'unusual' array types for now unless you feel strongly otherwise. Thanks again!
